Tests for next_occurrence break at certain times of day

## :white_check_mark: Checklist

<!--
Before reporting a problem please check the following. Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** develop @ 9e5b72a

## :package: Packages

<details>
<summary>List of packages</summary>

<!--
Powershell: Run `deployment/CheckRequirements.ps1` and paste the output below
Python: Run `hatch run pip freeze` and paste the output below
-->

```none
acme==2.6.0
annotated-types==0.6.0
appdirs==1.4.4
Arpeggio==2.0.2
attrs==23.2.0
azure-common==1.1.28
azure-core==1.30.1
azure-identity==1.16.0
azure-keyvault-certificates==4.8.0
azure-keyvault-keys==4.9.0
azure-keyvault-secrets==4.8.0
azure-mgmt-automation==1.0.0
azure-mgmt-compute==30.6.0
azure-mgmt-containerinstance==10.1.0
azure-mgmt-core==1.4.0
azure-mgmt-dns==8.1.0
azure-mgmt-keyvault==10.3.0
azure-mgmt-msi==7.0.0
azure-mgmt-network==25.3.0
azure-mgmt-rdbms==10.1.0
azure-mgmt-resource==23.0.1
azure-mgmt-storage==21.1.0
azure-storage-blob==12.19.1
azure-storage-file-datalake==12.14.0
azure-storage-file-share==12.15.0
certifi==2024.2.2
cffi==1.16.0
charset-normalizer==3.3.2
chevron==0.14.0
click==8.1.7
cryptography==42.0.6
-e git+ssh://git@github.com/jemrobinson/data-safe-haven.git@cb3d09511884bff82ab33b39e18b1db74dad7959#egg=data_safe_haven
dill==0.3.8
dnspython==2.4.2
fqdn==1.5.1
grpcio==1.60.1
idna==3.7
iniconfig==2.0.0
isodate==0.6.1
josepy==1.14.0
markdown-it-py==3.0.0
mdurl==0.1.2
msal==1.28.0
msal-extensions==1.1.0
msrest==0.7.1
oauthlib==3.2.2
packaging==24.0
parver==0.5
pluggy==1.5.0
portalocker==2.8.2
protobuf==4.25.3
psycopg==3.1.18
pulumi==3.115.0
pulumi_azure_native==2.39.0
pulumi_random==4.16.1
pycparser==2.22
pydantic==2.7.1
pydantic_core==2.18.2
Pygments==2.18.0
PyJWT==2.8.0
pyOpenSSL==24.1.0
pyRFC3339==1.1
pytest==8.2.0
pytest-asyncio==0.23.6
pytz==2023.4
PyYAML==6.0.1
requests==2.31.0
requests-oauthlib==2.0.0
rich==13.7.1
semver==2.13.0
setuptools==69.5.1
shellingham==1.5.4
simple-acme-dns==3.0.0
six==1.16.0
typer==0.12.3
typing_extensions==4.11.0
urllib3==2.2.1
validators==0.22.0
websocket-client==1.8.0
```

</details>

## :no_entry_sign: Describe the problem

<!--
What happens now and (if relevant) an explanation of why this is incorrect

Optionally include specific tasks (if any) in the order in which they need to be done.
- [ ] Task 1
- [ ] Task 2
- [ ] Task 3

Include a clear and concise description of what the behaviour should be when this issue is closed if this is not obvious.
-->

At certain times of day `next_occurrence` can choose a datetime more than 1 day in the future, causing the tests to break.

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
___________________ test_next_occurrence_is_within_next_day ____________________

    def test_next_occurrence_is_within_next_day():
        next_time = next_occurrence(5, 13, "Australia/Perth")
        dt_next_time = datetime.datetime.fromisoformat(next_time)
        dt_utc_now = datetime.datetime.now(datetime.UTC)
        assert dt_next_time > dt_utc_now
>       assert dt_next_time < dt_utc_now + datetime.timedelta(days=1)
E       AssertionError: assert datetime.datetime(2024, 5, 17, 21, 13, tzinfo=datetime.timezone.utc) < (datetime.datetime(2024, 5, 16, 16, 52, 28, 540053, tzinfo=datetime.timezone.utc) + datetime.timedelta(days=1))
E        +  where datetime.timedelta(days=1) = <class 'datetime.timedelta'>(days=1)
E        +    where <class 'datetime.timedelta'> = datetime.timedelta

tests/functions/test_strings.py:16: AssertionError
```

</details>

## :recycle: To reproduce

<!--
How can the current behaviour be reproduced? Include some steps to reproduce the problem.

Example:
- Deploy an SRE with option X set to Y
- Attempt to perform action Z
- Observe the previously mentioned failure/error message
-->
Run `hatch run test:test` at around 16:15 UTC.

--- a/data_safe_haven/functions/strings.py
+++ b/data_safe_haven/functions/strings.py
@@ -48,9 +48,11 @@ def next_occurrence(
             microsecond=0,
         )
         utc_dt = local_dt.astimezone(pytz.utc)
-        # Add one day until this datetime is in the future
-        utc_now = datetime.datetime.now(pytz.utc) + datetime.timedelta(minutes=1)
-        while utc_dt < utc_now:
+        # Add one day until this datetime is at least 1 hour in the future.
+        # This ensures that any Azure functions which depend on this datetime being in
+        # the future should treat it as valid.
+        utc_near_future = datetime.datetime.now(pytz.utc) + datetime.timedelta(hours=1)
+        while utc_dt < utc_near_future:
             utc_dt += datetime.timedelta(days=1)
         if time_format == "iso":
             return utc_dt.isoformat()

--- a/data_safe_haven/functions/strings.py
+++ b/data_safe_haven/functions/strings.py
@@ -46,8 +46,12 @@ def next_occurrence(
             minute=minute,
             second=0,
             microsecond=0,
-        ) + datetime.timedelta(days=1)
+        )
         utc_dt = local_dt.astimezone(pytz.utc)
+        # Add one day until this datetime is in the future
+        utc_now = datetime.datetime.now(pytz.utc) + datetime.timedelta(minutes=1)
+        while utc_dt < utc_now:
+            utc_dt += datetime.timedelta(days=1)
         if time_format == "iso":
             return utc_dt.isoformat()
         elif time_format == "iso_minute":

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -114,6 +114,7 @@ all = [
 [tool.hatch.envs.test]
 dependencies = [
   "coverage>=7.5.1",
+  "freezegun>=1.5",
   "pytest>=8.1",
   "pytest-mock>=3.14",
 ]

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -7,46 +7,52 @@ from data_safe_haven.exceptions import DataSafeHavenInputError
 from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
-@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
-def test_next_occurrence():
-    next_time = next_occurrence(5, 13, "Australia/Perth")
-    assert next_time == "2024-01-02T21:13:00+00:00"
+class TestNextOccurance:
+    @pytest.mark.parametrize("hour,minute,timezone,expected", [
+        (5, 13, "Australia/Perth", "2024-01-02T21:13:00+00:00"),
+        (0, 13, "Australia/Perth", "2024-01-02T16:13:00+00:00"),
+        (20, 13, "Australia/Perth", "2024-01-02T12:13:00+00:00"),
+        (20, 13, "Europe/London", "2024-01-02T20:13:00+00:00"),
+    ])
+    @freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
+    def test_next_occurrence(self, hour, minute, timezone, expected):
+        next_time = next_occurrence(hour, minute, timezone)
+        assert next_time == expected
 
+    @freeze_time(datetime.datetime(2024, 7, 2, 1, 0, tzinfo=datetime.UTC))
+    def test_dst(self):
+        next_time = next_occurrence(13, 5, "Europe/London")
+        assert next_time == "2024-07-02T12:05:00+00:00"
 
-@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
-def test_next_occurrence_timeformat():
-    next_time = next_occurrence(5, 13, "Australia/Perth", time_format="iso_minute")
-    assert next_time == "2024-01-02 21:13"
+    @freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
+    def test_timeformat(self):
+        next_time = next_occurrence(5, 13, "Australia/Perth", time_format="iso_minute")
+        assert next_time == "2024-01-02 21:13"
 
+    @freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.UTC))
+    def test_is_tomorrow(self):
+        next_time = next_occurrence(5, 13, "Australia/Perth")
+        assert next_time == "2024-01-03T21:13:00+00:00"
 
-@freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.UTC))
-def test_next_occurrence_is_tomorrow():
-    next_time = next_occurrence(5, 13, "Australia/Perth")
-    assert next_time == "2024-01-03T21:13:00+00:00"
+    def test_invalid_hour(self):
+        with pytest.raises(DataSafeHavenInputError) as exc_info:
+            next_occurrence(99, 13, "Europe/London")
+        assert exc_info.match(r"Time '99:13' was not recognised.")
 
+    def test_invalid_minute(self):
+        with pytest.raises(DataSafeHavenInputError) as exc_info:
+            next_occurrence(5, 99, "Europe/London")
+        assert exc_info.match(r"Time '5:99' was not recognised.")
 
-def test_next_occurrence_invalid_hour():
-    with pytest.raises(DataSafeHavenInputError) as exc_info:
-        next_occurrence(99, 13, "Europe/London")
-    assert exc_info.match(r"Time '99:13' was not recognised.")
+    def test_invalid_timezone(self):
+        with pytest.raises(DataSafeHavenInputError) as exc_info:
+            next_occurrence(5, 13, "Mars/OlympusMons")
+        assert exc_info.match(r"Timezone 'Mars/OlympusMons' was not recognised.")
 
-
-def test_next_occurrence_invalid_minute():
-    with pytest.raises(DataSafeHavenInputError) as exc_info:
-        next_occurrence(5, 99, "Europe/London")
-    assert exc_info.match(r"Time '5:99' was not recognised.")
-
-
-def test_next_occurrence_invalid_timezone():
-    with pytest.raises(DataSafeHavenInputError) as exc_info:
-        next_occurrence(5, 13, "Mars/OlympusMons")
-    assert exc_info.match(r"Timezone 'Mars/OlympusMons' was not recognised.")
-
-
-def test_next_occurrence_invalid_timeformat():
-    with pytest.raises(DataSafeHavenInputError) as exc_info:
-        next_occurrence(5, 13, "Australia/Perth", time_format="invalid")
-    assert exc_info.match(r"Time format 'invalid' was not recognised.")
+    def test_invalid_timeformat(self):
+        with pytest.raises(DataSafeHavenInputError) as exc_info:
+            next_occurrence(5, 13, "Australia/Perth", time_format="invalid")
+        assert exc_info.match(r"Time format 'invalid' was not recognised.")
 
 
 @pytest.mark.parametrize(

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -17,22 +17,22 @@ class TestNextOccurance:
             (20, 13, "Europe/London", "2024-01-02T20:13:00+00:00"),
         ],
     )
-    @freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
+    @freeze_time("1am on Jan 2nd, 2024")
     def test_next_occurrence(self, hour, minute, timezone, expected):
         next_time = next_occurrence(hour, minute, timezone)
         assert next_time == expected
 
-    @freeze_time(datetime.datetime(2024, 7, 2, 1, 0, tzinfo=datetime.UTC))
+    @freeze_time("1am on July 2nd, 2024")
     def test_dst(self):
         next_time = next_occurrence(13, 5, "Europe/London")
         assert next_time == "2024-07-02T12:05:00+00:00"
 
-    @freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
+    @freeze_time("1am on Jan 2nd, 2024")
     def test_timeformat(self):
         next_time = next_occurrence(5, 13, "Australia/Perth", time_format="iso_minute")
         assert next_time == "2024-01-02 21:13"
 
-    @freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.UTC))
+    @freeze_time("9pm on Jan 2nd, 2024")
     def test_is_tomorrow(self):
         next_time = next_occurrence(5, 13, "Australia/Perth")
         assert next_time == "2024-01-03T21:13:00+00:00"

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -7,19 +7,19 @@ from data_safe_haven.exceptions import DataSafeHavenInputError
 from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
-@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.timezone.utc))
+@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
 def test_next_occurrence():
     next_time = next_occurrence(5, 13, "Australia/Perth")
     assert next_time == "2024-01-02T21:13:00+00:00"
 
 
-@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.timezone.utc))
+@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
 def test_next_occurrence_timeformat():
     next_time = next_occurrence(5, 13, "Australia/Perth", time_format="iso_minute")
     assert next_time == "2024-01-02 21:13"
 
 
-@freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.timezone.utc))
+@freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.UTC))
 def test_next_occurrence_is_tomorrow():
     next_time = next_occurrence(5, 13, "Australia/Perth")
     assert next_time == "2024-01-03T21:13:00+00:00"

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -8,12 +8,15 @@ from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
 class TestNextOccurance:
-    @pytest.mark.parametrize("hour,minute,timezone,expected", [
-        (5, 13, "Australia/Perth", "2024-01-02T21:13:00+00:00"),
-        (0, 13, "Australia/Perth", "2024-01-02T16:13:00+00:00"),
-        (20, 13, "Australia/Perth", "2024-01-02T12:13:00+00:00"),
-        (20, 13, "Europe/London", "2024-01-02T20:13:00+00:00"),
-    ])
+    @pytest.mark.parametrize(
+        "hour,minute,timezone,expected",
+        [
+            (5, 13, "Australia/Perth", "2024-01-02T21:13:00+00:00"),
+            (0, 13, "Australia/Perth", "2024-01-02T16:13:00+00:00"),
+            (20, 13, "Australia/Perth", "2024-01-02T12:13:00+00:00"),
+            (20, 13, "Europe/London", "2024-01-02T20:13:00+00:00"),
+        ],
+    )
     @freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.UTC))
     def test_next_occurrence(self, hour, minute, timezone, expected):
         next_time = next_occurrence(hour, minute, timezone)

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -1,45 +1,28 @@
 import datetime
-import re
 
 import pytest
-import pytz
+from freezegun import freeze_time
 
 from data_safe_haven.exceptions import DataSafeHavenInputError
 from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
-def test_next_occurrence_will_occur_soon():
-    dt_utc_now = datetime.datetime.now(datetime.UTC)
+@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.timezone.utc))
+def test_next_occurrence():
     next_time = next_occurrence(5, 13, "Australia/Perth")
-    dt_utc_26_hours_time = dt_utc_now + datetime.timedelta(hours=26)
-    dt_next_time = datetime.datetime.fromisoformat(next_time)
-    assert dt_next_time > dt_utc_now
-    assert dt_next_time < dt_utc_26_hours_time
+    assert next_time == "2024-01-02T21:13:00+00:00"
 
 
-def test_next_occurrence_day_shift():
-    dt_utc_now = datetime.datetime.now(datetime.UTC)
-    dt_perth_now = dt_utc_now.astimezone(pytz.timezone("Australia/Perth"))
-    next_time = next_occurrence(
-        dt_perth_now.hour, dt_perth_now.minute, "Australia/Perth"
-    )
-    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, hours=2)
-    dt_next_time = datetime.datetime.fromisoformat(next_time)
-    assert dt_next_time > dt_utc_now
-    assert dt_next_time < dt_utc_tomorrow
-
-
-def test_next_occurrence_has_correct_time():
-    next_time = next_occurrence(5, 13, "Australia/Perth")
-    dt_as_utc = datetime.datetime.fromisoformat(next_time)
-    dt_as_local = dt_as_utc.astimezone(pytz.timezone("Australia/Perth"))
-    assert dt_as_local.hour == 5
-    assert dt_as_local.minute == 13
-
-
+@freeze_time(datetime.datetime(2024, 1, 2, 1, 0, tzinfo=datetime.timezone.utc))
 def test_next_occurrence_timeformat():
     next_time = next_occurrence(5, 13, "Australia/Perth", time_format="iso_minute")
-    assert re.match(r"\d\d\d\d-\d\d-\d\d \d\d:13", next_time)
+    assert next_time == "2024-01-02 21:13"
+
+
+@freeze_time(datetime.datetime(2024, 1, 2, 23, 0, tzinfo=datetime.timezone.utc))
+def test_next_occurrence_is_tomorrow():
+    next_time = next_occurrence(5, 13, "Australia/Perth")
+    assert next_time == "2024-01-03T21:13:00+00:00"
 
 
 def test_next_occurrence_invalid_hour():

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -17,6 +17,18 @@ def test_next_occurrence_will_occur_soon():
     assert dt_next_time < dt_utc_tomorrow
 
 
+def test_next_occurrence_day_shift():
+    dt_utc_now = datetime.datetime.now(datetime.UTC)
+    dt_perth_now = dt_utc_now.astimezone(pytz.timezone("Australia/Perth"))
+    next_time = next_occurrence(
+        dt_perth_now.hour, dt_perth_now.minute, "Australia/Perth"
+    )
+    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, hours=2)
+    dt_next_time = datetime.datetime.fromisoformat(next_time)
+    assert dt_next_time > dt_utc_now
+    assert dt_next_time < dt_utc_tomorrow
+
+
 def test_next_occurrence_has_correct_time():
     next_time = next_occurrence(5, 13, "Australia/Perth")
     dt_as_utc = datetime.datetime.fromisoformat(next_time)

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -11,10 +11,10 @@ from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 def test_next_occurrence_will_occur_soon():
     dt_utc_now = datetime.datetime.now(datetime.UTC)
     next_time = next_occurrence(5, 13, "Australia/Perth")
-    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, hours=2)
+    dt_utc_26_hours_time = dt_utc_now + datetime.timedelta(hours=26)
     dt_next_time = datetime.datetime.fromisoformat(next_time)
     assert dt_next_time > dt_utc_now
-    assert dt_next_time < dt_utc_tomorrow
+    assert dt_next_time < dt_utc_26_hours_time
 
 
 def test_next_occurrence_day_shift():

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -1,5 +1,3 @@
-import datetime
-
 import pytest
 from freezegun import freeze_time
 

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -8,10 +8,10 @@ from data_safe_haven.exceptions import DataSafeHavenInputError
 from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
-def test_next_occurrence_is_within_next_day():
+def test_next_occurrence_will_occur_soon():
     dt_utc_now = datetime.datetime.now(datetime.UTC)
     next_time = next_occurrence(5, 13, "Australia/Perth")
-    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, minutes=5)
+    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, hours=2)
     dt_next_time = datetime.datetime.fromisoformat(next_time)
     assert dt_next_time > dt_utc_now
     assert dt_next_time < dt_utc_tomorrow

--- a/tests/functions/test_strings.py
+++ b/tests/functions/test_strings.py
@@ -9,11 +9,12 @@ from data_safe_haven.functions import next_occurrence, sanitise_sre_name
 
 
 def test_next_occurrence_is_within_next_day():
-    next_time = next_occurrence(5, 13, "Australia/Perth")
-    dt_next_time = datetime.datetime.fromisoformat(next_time)
     dt_utc_now = datetime.datetime.now(datetime.UTC)
+    next_time = next_occurrence(5, 13, "Australia/Perth")
+    dt_utc_tomorrow = dt_utc_now + datetime.timedelta(days=1, minutes=5)
+    dt_next_time = datetime.datetime.fromisoformat(next_time)
     assert dt_next_time > dt_utc_now
-    assert dt_next_time < dt_utc_now + datetime.timedelta(days=1)
+    assert dt_next_time < dt_utc_tomorrow
 
 
 def test_next_occurrence_has_correct_time():

